toolchain/check/handle_interface.cpp

```diff
@@ -3,26 +3,154 @@
+
+      // TODO: Check that the generic parameter list agrees with the prior
+      // declaration.
+    } else {
+      // This is a redeclaration of something other than a interface.
+      context.DiagnoseDuplicateName(name_context.parse_node, existing_id);
+    }
+  }
+
+  // Create a new interface if this isn't a valid redeclaration.
+  if (!interface_decl.interface_id.is_valid()) {
+    // TODO: If this is an invalid redeclaration of a non-interface entity or
+    // there was an error in the qualifier, we will have lost track of the
+    // interface name here. We should keep track of it even if the name is
+    // invalid.
+    // TODO: should have a `Self` type id member
+    interface_decl.interface_id = context.interfaces().Add(
+        {.name_id =
+             name_context.state == DeclNameStack::NameContext::State::Unresolved
+                 ? name_context.unresolved_name_id
+                 : SemIR::NameId::Invalid,
+         .decl_id = interface_decl_id});
+  }
+
+  // Write the interface ID into the InterfaceDecl.
+  context.insts().Set(interface_decl_id, interface_decl);
+
+  return {interface_decl.interface_id, interface_decl_id};
 }
 
-auto HandleInterfaceDefinition(Context& context, Parse::NodeId parse_node)
+auto HandleInterfaceDecl(Context& context, Parse::NodeId /*parse_node*/)
     -> bool {
-  return context.TODO(parse_node, "HandleInterfaceDefinition");
+  BuildInterfaceDecl(context);
+  context.decl_name_stack().PopScope();
+  context.decl_state_stack().Pop(DeclState::Interface);
+  return true;
 }
 
 auto HandleInterfaceDefinitionStart(Context& context, Parse::NodeId parse_node)
     -> bool {
-  return context.TODO(parse_node, "HandleInterfaceDefinitionStart");
+  auto [interface_id, interface_decl_id] = BuildInterfaceDecl(context);
+  auto& interface_info = context.interfaces().Get(interface_id);
+
+  // Track that this declaration is the definition.
+  if (interface_info.definition_id.is_valid()) {
+    CARBON_DIAGNOSTIC(InterfaceRedefinition, Error,
+                      "Redefinition of interface {0}.", std::string);
+    CARBON_DIAGNOSTIC(InterfacePreviousDefinition, Note,
+                      "Previous definition was here.");
+    context.emitter()
+        .Build(parse_node, InterfaceRedefinition,
+               context.names().GetFormatted(interface_info.name_id).str())
+        .Note(context.insts().Get(interface_info.definition_id).parse_node(),
+              InterfacePreviousDefinition)
+        .Emit();
+  } else {
+    interface_info.definition_id = interface_decl_id;
+    interface_info.scope_id = context.name_scopes().Add();
+  }
+
+  // Enter the interface scope.
+  context.PushScope(interface_decl_id, interface_info.scope_id);
+
+  // TODO: Introduce `Self`.
+
+  context.inst_block_stack().Push();
+  context.node_stack().Push(parse_node, interface_id);
+  // TODO: do we need this for an interface?
+  context.args_type_info_stack().Push();
```

Comment:
Do we want to create a witness table for interfaces? (I think we might -- even if we don't use them at runtime, I think we'll probably want a compile-time representation for them.) If we do, I think this stack would be the most consistent place to build one.

If not -- or not for now -- then we can remove this and the pop below.

Reply:
I am not sure, but if so I think it would be in a later change.